### PR TITLE
Feature/capabilities endpoint

### LINKED ISSUE #5: Add capabilities endpoint

### PR DESCRIPTION
**Pull request by bartcunes**
Added the capabilities endpoint

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,8 +24,16 @@ docker pull ghcr.io/poort8/poort8.ishare.common:latest
...
     - CertificatePassword > The password of the iSHARE test certificate.
-    - CertificateChain > The certificates of the [iSHARE Test CA](https://dev.ishareworks.org/demo-and-testing/test-certificates.html#ishare-test-ca) as a comma separeted byte stream. Use step 2 to get these. In case your certificate is issued by _C=NL, O=iSHARE Foundation, CN=TEST iSHARE Foundation PKIoverheid Organisatie Server CA - G3_, you can use [this](https://raw.githubusercontent.com/POORT8/Poort8.Ishare.Common/master/ishare-test-ca-chain.txt) chain.
+    - CertificateChain > The certificates of the [iSHARE Test CA](https://dev.ishareworks.org/demo-and-testing/test-certificates.html#ishare-test-ca) as a comma separated byte stream. Use step 2 to get these. In case your certificate is issued by _C=NL, O=iSHARE Foundation, CN=TEST iSHARE Foundation PKIoverheid Organisatie Server CA - G3_, you can use [this](https://raw.githubusercontent.com/POORT8/Poort8.Ishare.Common/master/ishare-test-ca-chain.txt) chain.
     - CertificateChainPassword > Empty string for public test certificates: ""
+    - IshareRole > Your role in the iSHARE scheme, for example: Service Provider
```
Comment (mabvanaartrijk) [inline README.md]: Let's add the link to the iSHARE glossary: https://ishareworks.atlassian.net/wiki/spaces/IS/pages/70222187/Glossary

File: README.md [hunk 2]
```diff
@@ -24,8 +24,16 @@ docker pull ghcr.io/poort8/poort8.ishare.common:latest
...
+    - ApiVersion > The version of your iSHARE API
+    - TokenEndpointId > A self chosen GUID to identify your token endpoint. You can generate one on [GuidGen](https://www.guidgen.com)
+    - TokenEndpointUrl > The URL of your token endpoint. Usually your api address appended with /connect/token
```
Comment (mabvanaartrijk) [inline README.md]: Note that the container uses /token.

File: README.md [hunk 3]
```diff
@@ -24,8 +24,16 @@ docker pull ghcr.io/poort8/poort8.ishare.common:latest
...
+    - CapabilitiesEndpointId > A self chosen GUID to identify your capabilities endpoint. You can generate one on [GuidGen](https://www.guidgen.com)
+    - CapabilitiesEndpointUrl > The URL of your capabilities endpoint. Usually your api address appended with /capabilities
+    - PublicEndpoints > A comma separated string to represent all of your public iSHARE endpoints. The values of the endpoint itself are pipe separated, for example: GUID-public-endpoint-1|name-public-endpoint-1|description-public-endpoint-1|https://endpoints.myapi.com/public-endpoint-1;GUID-public-endpoint-2|name-public-endpoint-2|description-public-endpoint-2|https://endpoints.myapi.com/public-endpoint-2
```
Comment (mabvanaartrijk) [inline README.md]: As code formatting for readability?

File: README.md (hunk 1 above)
Comment (mabvanaartrijk) [reply README.md]: Should be one of the following values: Authorization Registry, Scheme Owner, Service Provider.

File: Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            }
+        }
+        return endpoints;
+    }
+
+    private IActionResult? HandleAuthorization(StringValues authorization, out string? audience)
+    {
+        audience = null;
+        if (string.IsNullOrEmpty(authorization)) return null;
+        if (authorization.Count != 1 || !authorization[0].StartsWith("Bearer "))
```
Comment (mabvanaartrijk) [inline Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs:81]: These checks are also in Core.

Comment (mabvanaartrijk) [reply Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs:81]: Ok, let's leave this in to pass the CTT.

File: Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            return new BadRequestObjectResult("Invalid token format");
```
Comment (mabvanaartrijk) [inline Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs]: Invalid token format: not a bearer token.

File: Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        var token = authorization[0].Replace("Bearer ", "");
+
+        try
+        {
+            _authenticationService.ValidateAuthorizationHeader(_configuration["ClientId"], authorization);
+        }
+        catch (Exception e)
+        {
+            _logger.LogWarning("Invalid authorization header, further checks needed. Message: {msg}", e.Message);
+            if (e.Message.StartsWith("IDX10223"))
```
Comment (mabvanaartrijk) [inline Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs]: We have a more validation errors.. And the generic error is confusing. Let's use a single unauthorized message: Invalid bearer token.

File: README.md (hunk 3 above)
Comment (mabvanaartrijk) [reply README.md]: comma separated >  semicolon in code

File: README.md [hunk 7]
```diff
@@ -24,8 +24,16 @@ docker pull ghcr.io/poort8/poort8.ishare.common:latest
...
+    - PrivateEndpoints > Like PublicEndpoints, a comma separated string to represent all of your private iSHARE endpoints, for example: GUID-private-endpoint-1|name-private-endpoint-1|description-private-endpoint-1|https://endpoints.myapi.com/private-endpoint-1;GUID-private-endpoint-2|name-private-endpoint-2|description-private-endpoint-2|https://endpoints.myapi.com/private-endpoint-2
```
Comment (mabvanaartrijk) [inline README.md]: comma separated > semicolon in code

File: Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        publicEndpoints.AddRange(RetrieveEndpointsFromConfig(_configuration["PublicEndpoints"]));
+        capabilitiesInfo.SupportedVersions.First().SupportedFeatures.Add(new PublicEndpoints(publicEndpoints));
+
+        if (audience is not null)
+        {
+            capabilitiesInfo.SupportedVersions.First().SupportedFeatures.Add(new RestrictedEndpoints(RetrieveEndpointsFromConfig(_configuration["PrivateEndpoints"])));
+        }
+        var additionalClaims = new List<Claim> { new Claim("capabilities_info", JsonSerializer.Serialize(capabilitiesInfo), JsonClaimValueTypes.Json) };
+
+        var token = _authenticationService.CreateTokenWithClaims(audience, additionalClaims);
```
Comment (mabvanaartrijk) [inline Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs]: var token = _authenticationService.CreateTokenWithClaims(**audience!**, additionalClaims); to get rid of the warning.

**Review by mabvanaartrijk — CHANGES_REQUESTED** [review]

File: Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs (hunk 8 above)
Comment (bartcunes) [reply Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs]: Fixed

File: README.md (hunk 7 above)
Comment (bartcunes) [reply README.md]: Changed

File: Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs (hunk 6 above)
Comment (bartcunes) [reply Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs]: Changed

File: Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs (hunk 5 above)
Comment (bartcunes) [reply Poort8.Ishare.Common/Capabilities/CapabilitiesController.cs]: Changed

File: README.md (hunk 1 above)
Comment (bartcunes) [reply README.md]: Changed

**Review by mabvanaartrijk — APPROVED** [review]

**Pull request merged**